src/types/decimal.rs

```diff
@@ -296,8 +296,42 @@ impl TryFrom<f64> for Decimal {
 
 impl Display for Decimal {
     fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        // TODO: This is correct, but not the most human-friendly format.
-        write!(f, "{}d{}", self.coefficient, self.exponent)
+        let digits = &*self.coefficient.magnitude.to_string();
+        let len = digits.len();
+        // The index of the decimal, relative to the magnitude representation
+        //       0123
+        // Given ABCDd-2, the decimal gets inserted at position 2.
+        let i_d = len as i64 + self.exponent;
+
+        if self.coefficient.sign == Sign::Negative {
+            write!(f, "-").unwrap();
+        };
+
+        if self.exponent == 0 && len > 6 {
+            // e.g. A.BCDEFGd6
+            write!(f, "{}.{}d{}", &digits[0..1], &digits[1..len], (i_d - 1))
+        } else if self.exponent == 0 {
+            // e.g. ABC.
+            write!(f, "{}.", &digits)
+        } else if self.exponent >= 0 {
+            // e.g. ABCd1
+            // let zeroes = "0".repeat(self.exponent as usize);
+            write!(f, "{}d{}", &digits, self.exponent)
+        } else {
+            // exponent < 0, there is a fractional component
+            if i_d > 0 {
+                // e.g. A.BC or AB.C
+                let ui_d = i_d as usize;
+                write!(f, "{}.{}", &digits[0..ui_d], &digits[ui_d..len])
+            } else if i_d > -6 {
+                // e.g. 0.ABC or 0.000ABC
+                let zeroes = "0".repeat(i_d.abs() as usize);
```

Comment:
This will allocate a new string, but you can achieve the same thing using the format string you pass to `write!`. See [the docs for _Fill/Alignment_ and _Sign/#/0_ for details](https://doc.rust-lang.org/std/fmt/#fillalignment).

Alternatively, create one very long `static ZEROS: &str = "0000000...";`  and just take a slice of the desired length 😛

Reply:
Using the format string passed to `write!` was a great suggestion, took a little fiddling but I like it :D